src/chan_vese.jl

```diff
@@ -0,0 +1,258 @@
+    m, n = last(CartesianIndices(𝚽ⁿ))[1], last(CartesianIndices(𝚽ⁿ))[2]
+    
+    while (del > tol) & (iter < max_iter)
+        ϵ = 1e-8
+        diff = 0
+
+        # Calculate the average intensities
+        @. H𝚽 = 𝚽ⁿ > 0 # Heaviside function
+        c₁, c₂ = calculate_averages(img, H𝚽, area, ∫u₀) # Compute c₁(𝚽ⁿ), c₂(𝚽ⁿ)
+
+        # Calculate the variation of level set 𝚽ⁿ
+        for idx in CartesianIndices(𝚽ⁿ) # Denote idx = (x, y)
+            # i₊ ≔ i₊(x, y), denotes 𝚽ⁿ(x, y + 1)'s CartesianIndex
+            # j₊ ≔ j₊(x, y), denotes 𝚽ⁿ(x + 1, y)'s CartesianIndex
+            # i₋ ≔ i₋(x, y), denotes 𝚽ⁿ(x, y - 1)'s CartesianIndex
+            # j₋ ≔ j₋(x, y), denotes 𝚽ⁿ(x - 1, y)'s CartesianIndex
+            # Taking notice that if 𝚽ⁿ(x, y) is the boundary of 𝚽ⁿ, than 𝚽ⁿ(x ± 1, y), 𝚽ⁿ(x, y ± 1) might be out of bound.
+            # So the pixel values of these outbounded terms are equal to 𝚽ⁿ(x, y)
+            i₊ = idx[2] != n ? idx + CartesianIndex(0, 1) : idx
+            j₊ = idx[1] != m ? idx + CartesianIndex(1, 0) : idx
+            i₋ = idx[2] != t ? idx - CartesianIndex(0, 1) : idx
+            j₋ = idx[1] != s ? idx - CartesianIndex(1, 0) : idx
+
+            𝚽₀  = 𝚽ⁿ[idx] # 𝚽ⁿ(x, y)
+            u₀ = img[idx] # u₀(x, y)
+            𝚽ᵢ₊ = 𝚽ⁿ[i₊] # 𝚽ⁿ(x, y + 1)
+            𝚽ⱼ₊ = 𝚽ⁿ[j₊] # 𝚽ⁿ(x + 1, y)
+            𝚽ᵢ₋ = 𝚽ⁿ[i₋] # 𝚽ⁿ(x, y - 1)
+            𝚽ⱼ₋ = 𝚽ⁿ[j₋] # 𝚽ⁿ(x - 1, y)
+
+            # Solve the PDE of equation 9 in paper[3]
+            C₁ = 1. / sqrt(ϵ + (𝚽ᵢ₊ - 𝚽₀)^2 + (𝚽ⱼ₊ - 𝚽ⱼ₋)^2 / 4.)
+            C₂ = 1. / sqrt(ϵ + (𝚽₀ - 𝚽ᵢ₋)^2 + (𝚽ⱼ₊ - 𝚽ⱼ₋)^2 / 4.)
+            C₃ = 1. / sqrt(ϵ + (𝚽ᵢ₊ - 𝚽ᵢ₋)^2 / 4. + (𝚽ⱼ₊ - 𝚽₀)^2)
+            C₄ = 1. / sqrt(ϵ + (𝚽ᵢ₊ - 𝚽ᵢ₋)^2 / 4. + (𝚽₀ - 𝚽ⱼ₋)^2)
+
+            K = 𝚽ᵢ₊ * C₁ + 𝚽ᵢ₋ * C₂ + 𝚽ⱼ₊ * C₃ + 𝚽ⱼ₋ * C₄
+            δₕ = h / (h^2 + 𝚽₀^2) # Regularised Dirac function
+            difference_from_average = - λ₁ * (u₀ - c₁) ^ 2 + λ₂ * (u₀ - c₂) ^ 2
+
+            𝚽ⁿ⁺¹[idx] = 𝚽 = (𝚽₀ + Δt * δₕ * (μ * K + difference_from_average)) / (1. + μ * Δt * δₕ * (C₁ + C₂ + C₃ + C₄))
+            diff += (𝚽 - 𝚽₀)^2
+        end
+
+        del = sqrt(diff / area)
+
+        if reinitial_flag
+            # Reinitialize 𝚽 to be the signed distance function to its zero level set
+            reinitialize(𝚽ⁿ⁺¹, 𝚽ⁿ, Δt, h)
+        else
+            𝚽ⁿ .= 𝚽ⁿ⁺¹
+        end
+  
+        iter += 1
+    end
+
+    return 𝚽ⁿ .> 0
+end
+
+function initial_level_set(shape::Tuple)
+    x₀ = reshape(collect(0:shape[begin]-1), shape[begin], 1)
+    y₀ = reshape(collect(0:shape[begin+1]-1), 1, shape[begin+1])
+    𝚽₀ = @. sin(pi / 5 * x₀) * sin(pi / 5 * y₀)
+end
+
+function calculate_averages(img::AbstractArray{T, N}, H𝚽::AbstractArray{S, N}, area::Int64, ∫u₀::Float64) where {T<:Real, S<:Bool, N}
+    ∫u₀H𝚽 = 0
+    ∫H𝚽 = 0
+    for i in eachindex(img)
+        if H𝚽[i]
+            ∫u₀H𝚽 += img[i]
+            ∫H𝚽 += 1
+        end
+    end
```

Comment:
```suggestion
    @inbounds @simd for i in eachindex(img)
        if H𝚽[i]
            ∫u₀H𝚽 += img[i]
            ∫H𝚽 += 1
        end
    end
```

`@simd` might not work here as there is an if-branch inside the loop.

Reply:
The `@inbounds` is a good idea, but yes, the `@simd` won't do anything (best not to use it in such cases).